### DISCUSSION TITLE
Upcoming releases: v6.10 & v7

### DISCUSSION THREAD

**Discussion by nfriedly** [nfriedly/express-rate-limit#372] [Ideas]
I have a few things in mind for the v7 release, some of which would benefit from warning users of ahead of time:

### v6.10.0 - done and shipped!

- [x] Warn for use of `onLimitReached`
  * PR: https://github.com/express-rate-limit/express-rate-limit/pull/373
- [x] Warn for use of `max=0`
  * PR: https://github.com/express-rate-limit/express-rate-limit/pull/370
  * Previous discussion: https://github.com/express-rate-limit/express-rate-limit/discussions/369
  * consider modifying the validator to let this check run indefinitely, or at least for more than the first request
- [x] Add support for draft-7 standard headers
  * PR: https://github.com/express-rate-limit/express-rate-limit/pull/376
- [x] Warn for `draft_polli_ratelimit_headers`
   * PR: https://github.com/express-rate-limit/express-rate-limit/pull/377
- [x] readme: use named import for both cjs and esm
  * This should help to avoid https://github.com/express-rate-limit/express-rate-limit/issues/360
  * Commit: https://github.com/express-rate-limit/express-rate-limit/commit/9ba2ed2903cec624453a5215a126df06e8bec6bb
- [x] Add a new validation check that hits is a positive number (https://github.com/express-rate-limit/express-rate-limit/issues/353)
  * ~~Also, fix the root cause in the Memcached store~~ 
    * Not really a bug, they just tried something undocumented and it failed in an unexpected way. However, it works when following the documentation.  A validation check is probably sufficient.
  * PR: https://github.com/express-rate-limit/express-rate-limit/pull/379

### v7.0.0 - done and shipped!

- [x] change `max=0` behavior to just call `handler` on every request
  - update warning to indicate that the change has now happened
  - PR: https://github.com/express-rate-limit/express-rate-limit/pull/380
- [x] Drop support for `draft_polli_ratelimit_headers` option
   - update the warning
- [x] drop support for `onLimitReached` 
    - update the validation error
    - PR: https://github.com/express-rate-limit/express-rate-limit/pull/382
- [x] upgrade dts-bundle-generator and typescript
- [x] Bump min node.js version to node 16.
    - Related: bump the typescript target to ES2022 per https://stackoverflow.com/questions/67371787/what-typescript-configuration-produces-output-closest-to-node-js-16-capabilities/67371788#67371788
    - PR: https://github.com/express-rate-limit/express-rate-limit/pull/383
- [x] Updated MemoryStore
    - PR: https://github.com/express-rate-limit/express-rate-limit/pull/378
- [x] Deprecate `max`, replace it with `limit`. 
    - Continue to support both without a warning for now.
    - PR: https://github.com/express-rate-limit/express-rate-limit/pull/381
- [x] Consider making validations more flexible - allow some to run on more than one request, allow them to be individually disabled 
  - PR for individual disabling: https://github.com/express-rate-limit/express-rate-limit/pull/392
- [x] Rename `req.rateLimit.current` to `req.rateLimit.used`
  - PR: https://github.com/express-rate-limit/express-rate-limit/pull/391

#### Headers
Given that the standard headers draft is still undergoing significant changes, I think we should skip most of the warnings around headers (except the `draft_polli_ratelimit_headers` option), and *definitely* not change the defaults.

##### v6.10:
- ~~Warn for v5 and earlier `headers` option~~
- ~~Warn for unset headers options~~

##### v7
-  ~~change headers defaults to `standardHeaders: true, legacyHeaders: false`~~
  - ~~update the warning for now, probably remove it eventually~~
  - ~~Maybe add a warning for `standardHeaders: true`?~~
- ~~Drop support for v5 headers options~~

#### Other changes omitted from v7:
- ~~maybe suppress `max=0` warning if a custom handler is set (this will be the case for express-slow-down and could conceivably be helpful elsewhere)~~
      - the ability to disable specific validations mostly obviates the need for this
- ~~Maybe: switch esbuild to tsc to use decorators?~~
  - not needed, I worked out a solution to disable specific validators without needing this.

**Comment by gamemaker1** [comment]
Should we also drop node 14? It's been [EOL](https://endoflife.date/nodejs) for almost 2 years, and active security support ended 3 months ago.

> change headers defaults to standardHeaders: true, legacyHeaders: false

We should also remove the `draft_polli` option.

> switch esbuild to tsc to use decorators? 

We could try out tsc for the decorators, if it makes writing validations easier.

**Reply by nfriedly** [reply]
> Should we also drop node 14? It's been [EOL](https://endoflife.date/nodejs) for almost 2 years, and active security support ended 3 months ago.

Yeah, seems like a good idea, especially if it's making testing more difficult.

> We should also remove the draft_polli option.

Yes, and we should add a warning for that now too.

>  > switch esbuild to tsc to use decorators?
> 
> We could try out tsc for the decorators, if it makes writing validations easier

Yeah, I'm not 100% on this one, but I might play with it some.

I'll update the top post with the other couple of things.

**Comment by gamemaker1** [comment]
I was wondering if we should drop the default export too - and only have a named export. It'll allow us to get rid of the [ugly footer we tape on](https://github.com/express-rate-limit/express-rate-limit/blob/main/package.json#L59) to the end of the cjs files.

**Reply by nfriedly** [reply]
No, I think that'd be too big of a breaking change. The changes I want to make shouldn't affect anyone who's been following our guidelines since the V6 release, but removing the default export would break basically everyone who followed our instructions up to this point.

I suppose we could deprecate it in V7 for removal in an eventual V8 release, but even that feels like too much. I think we're stuck with the ugly footer for now.

Honestly, I'd be more inclined to deprecate the cjs release entirely, and switch to pure esm. But I'm not sure if it's too soon for that.

I kind of wish we had some analytics. But, even making that opt-in probably wouldn't fly well with some folks. And we'd only get usage data for releases after we added it, anyways.

**Reply by gamemaker1** [reply]
I think we should deprecate it in v8, to set up the path for an eventual ESM-only release. Although, as you said, that might take a lot more time - maybe even a couple of years.

> I kind of wish we had some analytics. But, even making that opt-in probably wouldn't fly well with some folks. And we'd only get usage data for releases after we added it, anyways.

We could try to parse through the [github repos that use our package](https://github.com/express-rate-limit/express-rate-limit/network/dependents) using a script, that might give us a hint.

**Reply by nfriedly** [reply]
Hmm.. although if we switch to pure esm, other esm projects could still use a default import, it's only cjs projects that would need to change. Still feels like too big of a breaking change, though.

I do like the idea of parsing through all the projects that depend on erl!

**Comment by nfriedly** [comment]
Hum, it looks like the headers standardization went through a major change: https://author-tools.ietf.org/iddiff?url1=draft-ietf-httpapi-ratelimit-headers-06&url2=draft-ietf-httpapi-ratelimit-headers-07&difftype=--html

Instead of separate headers for limit, remaining, and reset, they're now fields in a single header. 

I think we'll want to support this, but now I'm not sure about making it the default since it's still changing. I also think we want some way to request the behavior from the earlier draft. I don't like `true` pointing to an older version, but I also don't like changing the behavior under everyone's feet. Maybe we'll require it be a value like `draft-6` or `draft-7` and warn, but use `draft-6` if it's set to `true`?

**Reply by gamemaker1** [reply]
> Maybe we'll require it be a value like draft-6 or draft-7 and warn, but use draft-6 if it's set to true?

That sounds like a good idea.

Also, I noticed that there was a `RateLimit-Policy` header in the draft that we were not setting earlier, should we set that one too?

**Reply by nfriedly** [reply]
Yeah, we'll add `RateLimit-Policy` for the `draft-7` mode.

I'm a little annoyed at myself for calling it `standardHeaders` when it's still a changing draft :/

On the upside, going this route is not a breaking change, so we can ship it in 6.10.

I should probably read all the drafts carefully and see if there are any other details that we need to take into account.

**Reply by nfriedly** [reply]
Actually, I think we should add support for `draft-6` and `draft-7`, and update the readme in 6.10.

But I don't think we should start warning about `true` until 7.0. I feel like it's just too annoying otherwise.

**Reply by gamemaker1** [reply]
> Yeah, we'll add RateLimit-Policy for the draft-7 mode. On the upside, going this route is not a breaking change, so we can ship it in 6.10.

Cool! Should we change the default to `draft-7` and start warning at v7 then?

> I'm a little annoyed at myself for calling it standardHeaders when it's still a changing draft :/

Same :/

**Reply by nfriedly** [reply]
> Cool! Should we change the default to draft-7 and start warning at v7 then?

No, I think we should keep the behavior of `true` the same for v7, but just warn about it. 

I don't think we should change its behavior until the draft is fully standardized. Then we'll do another Semver major release (v8 or whatever)  where we make `true`  activate whatever the final standardized behavior is.

**Reply by nfriedly** [reply]
Now that the rate limit header is getting more complex, maybe we should write a parser? Probably as a separate package though.

**Reply by gamemaker1** [reply]
> No, I think we should keep the behavior of true the same for v7, but just warn about it. I don't think we should change its behavior until the draft is fully standardized. Then we'll do another Semver major release (v8 or whatever) where we make true activate whatever the final standardized behavior is.

Alright!

**Reply by gamemaker1** [reply]
> Now that the rate limit header is getting more complex, maybe we should write a parser? Probably as a separate package though.

Sure, that sounds like a good idea.

**Reply by nfriedly** [reply]
I did some initial work on a headers parser. It works for the couple of happy-cases tests I wrote, but it still needs some work before it's ready to ship: https://github.com/express-rate-limit/ratelimit-header-parser

**Reply by gamemaker1** [reply]
Oooh nice! I'll help with the todos.

**Comment by nfriedly** [comment]
I just had a thought: for deprecation notices, we've been using `WRN` rather than `ERR` in the codes. That makes sense right now, but are we going to want to make them errors rather than warnings in 7.x? Perhaps the codes should just be `ERR` from the start then?

**Reply by gamemaker1** [reply]
I think deprecation notices should be only warnings, not errors.

Even when we remove the options in v7, the notices are still _warning_ users that expected behaviour might not be seen, because they used the wrong option.

**Comment by nfriedly** [comment]
One more change I think we should make is to rename the `max` option to `limit` for consistency.  But I think we'll have to keep `max` around as a deprecated option indefinitely, so I don't think it needs a validation check. Not technically a breaking change, but probably a good fit for 7.x anyways.

**Reply by gamemaker1** [reply]
Alright!

**Comment by nfriedly** [comment]
Last PR for 6.10 is up: https://github.com/express-rate-limit/express-rate-limit/pull/379

**Reply by gamemaker1** [reply]
Merged! Waiting on #376, let's do a release once that is merged?

**Comment by gamemaker1** [comment]
Kind of a note to myself because I thought of this twice today and came to the same conclusion both times:

Can't use `pkgroll` because it outputs CJS with the following exports -

```js
exports.MemoryStore = MemoryStore;
exports["default"] = rateLimit;
exports.rateLimit = rateLimit;
```

So anyone using:

```js
const rateLimit = require('express-rate-limit')
```

will hit an error that says `rateLimit` is not a function. Instead, they would need to use:

```js
const rateLimit = require('express-rate-limit').default
```

Ppl using typescript with [`esModuleInterop: true`](https://www.typescriptlang.org/tsconfig#esModuleInterop) or [`allowSyntheticDefaultImports: true`](https://www.typescriptlang.org/tsconfig#allowSyntheticDefaultImports) won't need to make this change, since typescript [does it for them](https://stackoverflow.com/a/56348146).

However, I feel it would be quite presumptous if we forced JS users to change their imports just so we could make bundling the library easier for us. Instead, we could shift to `pkgroll` when we switch to only named exports, which is the recommended method of exporting things in a dual-cjs-esm package.

---

The benefit of switching to `pkgroll` is that all three of our build commands - `build:cjs,esm,types` - get replaced with a single line: `pkgroll --target es2022 --src source/`, and we don't need to worry about issues like #361 again.

**Comment by nfriedly** [comment]
Hey, this is kind of unrelated, but did something change with our jest setup in the past few weeks? I used to be able to run individual tests from within vs code, and now that seems broken.

**Reply by gamemaker1** [reply]
I did move the jest configuration to a separate file - `config/jest.json`, maybe VSCode hasn't detected that automatically?

**Reply by nfriedly** [reply]
Yeah, that might be it. Was it in package.json before?

**Reply by gamemaker1** [reply]
Yea

**Reply by gamemaker1** [reply]
I was actually planning to move all package.json config into separate config files, for `jest` and `xo` and `lint-staged`.

**Reply by nfriedly** [reply]
https://github.com/express-rate-limit/express-rate-limit/pull/387 fixes jest testing in vs code. It's still in a separate config file, but at the top-level now.

**Comment by nfriedly** [comment]
What do you think about changing `req.rateLimit.current` to `req.rateLimit.used` in v7 to better align with the terminology in the IETF draft?

We could keep the old one around possibly with `Object.defineProperty()` to make it not show up, e.g.

Object.defineProperty(rateLimitInfo, 'current', { configurable: false, enumerable: false, get: function() { return this.used } })

That would allow code calling `req.rateLimit.current` to continue working, but hide it from iteration and `JSON.stringify()` and the like. We could even throw a validation warning in the getter if we wanted. (Although, unfortunately, that would fire in the cases where it doesn't break anything, and *not* in the cases where it does, so maybe not worth it.)

**Reply by gamemaker1** [reply]
Sounds good!

**Reply by gamemaker1** [reply]
Do you want to put this in v7 or v6.11?

**Reply by nfriedly** [reply]
Probably v7, because it is a breaking change in some scenarios.

**Comment by nfriedly** [comment]
Ok, there's only two items left on the v7 checklist, and I think we can skip them both:

> *  maybe suppress max=0 warning if a custom handler is set (this will be the case for express-slow-down and could conceivably be helpful elsewhere)

Now that specific validations can be disabled by configuration, I don't think this is as important.

> * Maybe: switch esbuild to tsc to use decorators? 

This was primarily to make the validations changes I had in mind easier, but I ended up working them out without this. So I think we can skip it and stick with what works.

I think we're ready, what do you think?

**Reply by gamemaker1** [reply]
:shipit: :rocket:

**Comment by gamemaker1** [comment]
Done! Yay!
